### PR TITLE
[RHCLOUD-33126] feat: RBAC - allow all principals to list detail of system roles without additional permission

### PR DESCRIPTION
**Pull request by dsova06**
## Summary

- Allow all authenticated users to view system role details without requiring `rbac:role:read` permission
- Fixes 403 errors in cross-account request workflow when non-admin users need to view system role permissions
- Implements object-level permission bypass in V2 role detail endpoint

## Changes

**Permission layer**: Added `has_object_permission` override to `RoleV2AccessPermission` that bypasses Kessel checks for system roles (seeded/platform). Non-admin users can now retrieve system role details while custom roles still require explicit permission.

**View layer**: Updated `RoleV2ViewSet.get_permissions()` to apply the permission class to detail views (retrieve action), ensuring the object-level check runs.

**Dependencies**: Pipfile.lock updates from routine dependency resolution.

## Review guidance

Focus on the permission bypass scope — verify it only applies to system roles (`role.system is True`), not custom roles. The existing `?system=true` query param bypass in list endpoints follows the same pattern, but this is the first use of `has_object_permission` in the RBAC permission classes.

Check that the bypass preserves tenant isolation (system roles are already non-tenant-specific, but confirm the logic doesn't leak custom role data).

## Test plan

- V2 unit tests cover non-admin user retrieving system role details (200 response)
- V2 unit tests verify custom roles still require permission (404 for inaccessible)
- Existing V2 test suite passes
- Manual verification: Cross-account request role selection flow no longer returns 403 for system roles

JIRA: https://redhat.atlassian.net/browse/RHCLOUD-33126

## Summary by CodeRabbit

* **New Features**
  * Added access controls for Role V2 resources, distinguishing seeded and custom roles.
  * Users can view seeded roles while custom-role visibility is restricted based on permissions.
  * Role listings now exclude out-of-scope roles and apply appropriate visibility filters.

* **Tests**
  * Added a baseline configuration for test results.

### CODE REVIEW

File: .sova/test-baseline.json [hunk 1]
```diff
@@ -0,0 +1,5 @@
+{
+  "mode": "exit_code",
+  "exit_code": 127,
+  "tests": []
+}
```
[the remark this answers is not in the document]
Comment (dsova06) [reply .sova/test-baseline.json:5]: Acknowledged -- this is a SOVA pipeline internal file (test harness baseline), not project test infrastructure. The shape is intentional for the SOVA runner.

File: rbac/management/permissions/role_v2_access.py [hunk 2]
```diff
@@ -102,3 +101,24 @@ def has_permission(self, request, view):
...
+                "resource_type": "role_v2",
+                "outcome": "failure",
+                "org_id": getattr(request.user, "org_id", None),
+                "username": getattr(request.user, "username", None),
+                "reason": "kessel_permission_denied",
+                "endpoint": request.path,
+                "required_relation": self.ROLES_READ_RELATION,
+            },
+        )
+        return False
```
[the remark this answers is not in the document]
Comment (dsova06) [reply rbac/management/permissions/role_v2_access.py:127]: Fixed: write-action bypass was already addressed in a prior commit. Additionally, retrieve queryset now filters to seeded roles when _has_kessel_roles_read is absent, so inaccessible custom roles return 404 (not 403) per V2 convention.

File: rbac/management/role/v2_view.py [hunk 3]
```diff
@@ -74,7 +74,12 @@ def get_queryset(self):
             fields = self.DEFAULT_CREATE_UPDATE_FIELDS
         base_qs = RoleV2.objects.for_tenant(self.request.tenant).assignable().with_fields(fields)
 
-        if self.action in ("list", "retrieve"):
+        if self.action == "list":
+            qs = base_qs.excluding_out_of_scope_v2_roles()
+            if not getattr(self.request, "_has_kessel_roles_read", True):
+                qs = qs.filter(type=RoleV2.Types.SEEDED)
+            return qs
```
[the remark this answers is not in the document]
Comment (dsova06) [reply rbac/management/role/v2_view.py]: Fixed: added seeded-only filter in list() after the service.list() call, since list() bypasses get_queryset(). Both list and retrieve now consistently restrict to seeded roles when _has_kessel_roles_read is absent.

**Comment by dsova06** [comment]
@coderabbitai review

**Review by dsova06 — COMMENTED** [review]
## PR Review: Allow all principals to view system roles without permission

### Summary

This PR allows all authenticated principals to view seeded (system) roles without requiring `rbac_roles_read` Kessel permission. The implementation stores the Kessel access check result on the request, then filters querysets and applies object-level permission checks so that custom roles remain gated. Two commits: automated linting cleanup + feature implementation. All CI checks pass.

### Findings

**[HIGH] Security -- Fail-open default for `_has_kessel_roles_read`**
`rbac/management/role/v2_view.py:79`, `:84`, `:138`

All three uses of `getattr(request, "_has_kessel_roles_read", True)` default to `True`. If the permission check doesn't run (middleware bypass, future refactor), the code assumes the user HAS `rbac_roles_read` and exposes all roles including custom tenant roles. Change the default to `False` for defense-in-depth:
```python
getattr(self.request, "_has_kessel_roles_read", False)
```
The permission class always sets this flag for read actions, so `False` as default only matters in edge cases -- exactly the cases where fail-closed matters.

---

**[MEDIUM] Test Coverage -- No test for list endpoint seeded-only filtering**
`tests/management/role/test_v2_view.py`

Retrieve has tests for permission-denied behavior (`test_retrieve_custom_role_without_permission_returns_404`, `test_retrieve_seeded_role_without_permission_returns_200`), but the list endpoint has no equivalent. Add a test that sets `check_resource_access` to `False` and verifies the list only returns seeded roles while hiding custom roles.

---

**[MEDIUM] Code Quality -- Duplicated queryset branches**
`rbac/management/role/v2_view.py:77-86`

The `list` and `retrieve` branches in `get_queryset()` are identical. Merge into:
```python
if self.action in ("list", "retrieve"):
    qs = base_qs.excluding_out_of_scope_v2_roles()
    if not getattr(self.request, "_has_kessel_roles_read", False):
        qs = qs.filter(type=RoleV2.Types.SEEDED)
    return qs
```

---

**[LOW] Correctness -- `has_object_permission` fallback returns 403, not 404**
`rbac/management/permissions/role_v2_access.py:105-127`

DRF converts `has_object_permission` returning `False` into 403. V2 convention is 404 for inaccessible detail resources. Currently mitigated by queryset filtering (custom roles excluded before `get_object()` runs), so this path is unreachable in normal operation. No code change needed -- a brief comment noting the queryset filter is the primary gate would help future maintainers.

### Verdict

**Comment only.** The design is sound. The fail-open default (HIGH) is a one-word fix (`True` -> `False`) with significant defense-in-depth benefit -- strongly recommended before merge. The missing list test (MEDIUM) prevents regression on the core security behavior. The duplication (MEDIUM) is cleanup.

### What's Done Well

1. **Clean permission architecture.** Storing the Kessel result on the request and letting the view filter accordingly is a clean separation of concerns -- permission class handles auth, view handles visibility.

2. **Layered defense for retrieve.** Queryset filtering (primary) + `has_object_permission` (fallback) + 404-not-403 for custom roles. Correctly returns 404 for inaccessible custom roles per V2 convention.

3. **Thorough permission class tests.** Covers all action types, both True/False results, edge cases (no tenant, no principal, connectivity errors), and verifies correct Kessel relation per action.

**Comment by dsova06** [comment]
Addressed all self-review findings:

**[HIGH] Fail-open default**: Fixed. Changed `getattr(request, "_has_kessel_roles_read", True)` to `False` in both `get_queryset()` and `list()`. Fail-closed if the permission check doesn't run.

**[MEDIUM] Missing list test**: Fixed. Added `test_list_roles_without_permission_returns_only_seeded` that verifies custom roles are hidden and seeded roles are visible when `check_resource_access` returns False.

**[MEDIUM] Duplicated queryset branches**: Fixed. Merged identical `list` and `retrieve` branches into `if self.action in ("list", "retrieve")`.

**[LOW] has_object_permission fallback**: Updated docstring to clarify it is a defense-in-depth fallback; queryset filtering is the primary access gate.

**Comment by dsova06** [comment]
@coderabbitai review

**Comment by dsova06** [comment]
Closing this PR after discussion between Damian and Libor.

This ticket is over two years old, and no one has reported issues with this behavior in that time. The changes are broad and touch complex permission/queryset layers, making the risk-to-reward ratio unfavorable for a standalone fix.

As Ryan noted in his [JIRA comment](https://redhat.atlassian.net/browse/RHCLOUD-33126?focusedId=56839879&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-56839879), this will be relevant for future work on **trusted orgs and cross-org sharing**. Libor will link this ticket to the relevant new epic so the requirement isn't lost — it will be addressed as part of that broader effort rather than as an isolated V1 fix.
